tools/crm_mon.c

```diff
@@ -2807,18 +2853,363 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
+                        /* damn single linked list */
+                        free(hp->target);
+                        free(hp->action);
+                        free(np->origin);
+                        np->origin = hp->origin;
+                        free(np->delegate);
+                        np->delegate = hp->delegate;
+                        free(np->client);
+                        np->client = hp->client;
+                        np->completed = hp->completed;
+                        tmp = hp;
+                        hp = hp->next;
+                        free(tmp);
+                        break;
+                    }
+                }
+                if (np->next) {
+                    continue;
+                }
+            }
+            np = 0; /* let outer loop progress hp */
+            break;
+        }
+        /* simply move hp from history to new */
+        if (np == NULL) {
+            tmp = hp->next;
+            hp->next = new;
+            new = hp;
+            hp = tmp;
+        }
+    }
+    return new;
+}
+
+/*!
+ * \internal
+ * \brief Sort the stonith-history
+ *        sort by competed most current on the top
+ *        pending actions lacking a completed-stamp are gathered at the top
+ *
+ * \param[in] history    List of stonith actions
+ *
+ */
+static stonith_history_t *
+sort_stonith_history(stonith_history_t *history)
+{
+    stonith_history_t *new = NULL, *pending = NULL, *hp, *np, *tmp;
+
+    for (hp = history; hp; ) {
+        tmp = hp->next;
+        if ((hp->state == st_done) || (hp->state == st_failed)) {
+            /* sort into new */
+            if ((!new) || (hp->completed > new->completed)) {
+                hp->next = new;
+                new = hp;
+            } else {
+                np = new;
+                do {
+                    if ((!np->next) || (hp->completed > np->next->completed)) {
+                        hp->next = np->next;
+                        np->next = hp;
+                        break;
+                    }
+                    np = np->next;
+                } while (1);
+            }
+        } else {
+            /* put into pending */
+            hp->next = pending;
+            pending = hp;
+        }
+        hp = tmp;
+    }
+
+    /* pending actions don't have a completed-stamp so make them go front */
+    if (pending) {
+        stonith_history_t *last_pending = pending;
+
+        while (last_pending->next) {
+            last_pending = last_pending->next;
+        }
+
+        last_pending->next = new;
+        new = pending;
+    }
+    return new;
+}
+
+/*!
+ * \internal
+ * \brief Turn stonith action into a better readable string
+ *
+ * \param[in] action     Stonith action
+ */
+static char *
+fence_action_str(const char *action)
+{
+    char *str = NULL;
+
+    if (action == NULL) {
+        str = strdup("unknown");
+    } else if (action[0] == 'o') { // on, off
+        str = crm_concat("turn", action, ' ');
```

Comment:
Maybe "turning" e.g. "turning on of node1 succeeded"
and/or translate "on" to "unfencing"